File: BST/tree.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>  

#define VERBOSE 1

namespace tree {

    class Node {
        int key;
        int value;
        Node* left;
        Node* right;
    public:
        Node(int k, int v);
        int getVal();
        int getKey();
        Node* getLeft();
        Node* getRight();
        void put(int k, int v);
        Node* min();
        Node* pop(int k);
        void traverse();
        void reset();
    };

    Node::Node(int k, int v) {
        this->key = k;
        this->value = v;
        this->left = NULL;
        this->right = NULL;
    }

    int Node::getVal() {
        return this->value;
    }

    int Node::getKey() {
        return this->key;
    }

    Node* Node::getLeft() {
        return this->left;
    }

    Node* Node::getRight() {
        return this->right;
    }

    void Node::put(int k, int v) {
        if (k == this->key) {
            this->value = v;
            return;
        }
        if (k < this->key) {
            if (this->left == NULL) {
                this->left = new Node(k, v);
                return;
            }
            this->left->put(k, v);
        }
        else {
            if (this->right == NULL) {
                this->right = new Node(k, v);
                return;
            }
            this->right->put(k, v);
        }
    }

    Node* Node::min() {
        if (this == NULL) return NULL;
        if (this->left == NULL) return this;
        else this->left->min();
    }

    /* 
     * pop(k) - remove the node with key = k
     * Example) (1-3-4)-6-(7-8-9)
     *       6
     *    3     8
     *   1 4   7 9
     * 
     * root = root->pop(8)  
     *
     * this[6].pop(8)
     *  (k = 8 > this[6]->key) => this[6]->right = this[6]->right->pop(8) = (2) (== [8]->pop(8))
     * 
     *      (k = 8 == this[8]->key)
     *      (this[8]->left  != NULL)
     *      (this[8]->right != NULL)     
     *      tem1            = [8]->right (== [9])
     *      this[8]->key    = [9]->key (== 9)
     *      this[9']->right = this[9']->right->pop(tem1->right (== 9)) = (1)
     * 
     *          (k = 9 == [9]-> key)
     *          ([9]->left  == NULL)
     *          tem2    = [9]->right = NULL
     *          delete [9]
     *          return NULL (1)
     * 
     *      this[9']->right = (1) = NULL
     *      return this[9'] = (2)
     *  
     *  this[6]->right = (2) = [9']
     *  return this[6]
     * 
     * root = [6]
     * 
     * Result: 
     *       6
     *    3     9'
     *   1 4   7 -
     */
    Node* Node::pop(int k) {
        if (this == NULL) return this;
        if (k < this->key) this->left = this->left->pop(k); 
        else if (k > this->key) this->right = this->right->pop(k);
        else {
            if (this->left == NULL) {
                Node* temp = this->right;
                delete this;
                return temp;
            }
            else if (this->right == NULL) {
                Node* temp = this->left;
                delete this;
                return temp;
            }
            else {
                Node* temp = this->right->min();        // successor
                this->key = temp->key;
                this->right = this->right->pop(temp->key);
            }
        }
        return this;
    }

    void Node::traverse() {
        if (this == NULL) {
            printf(" empty tree ");
            return;
        }
        if (this->left != NULL) {
            printf("(");
            this->left->traverse();
            printf(")-");
        }
        printf("%d", this->key);
        if (this->right != NULL) {
            printf("-(");
            this->right->traverse();
            printf(")");
        }
    }

    void Node::reset() {
        if (this == NULL) {
            delete this;
        }
        if (this->left != NULL) {
            this->left->reset();
        }
        if (this->right != NULL) {
            this->right->reset();
        }
        delete this;
    }
}

void line() {
    printf("--------------------------------------------------------------------\n");
}

int main() {
    line();
    printf("Binary Search Tree in C++\n");
    using namespace tree;
    Node* root;
    int status = 1;
    int init = 1;
    while (status) {
        char cmd[20];
        int key = 0;
        printf(">> enter a command: ");
        scanf("%19s", cmd);
        if (!strcmp(cmd, "exit")) 
        {
            root->reset();
            delete root;
            status = 0;
        }
        else if (!strcmp(cmd, "put")) {
            printf(">> enter an integer to put: ");
            scanf("%d", &key);
            if (init == 1) {
                root = new Node(key, key);
                init = 0;
            }
            else
            {
                root->put(key, key);
            }
            printf(">> ");
            root->traverse();
            printf(" \n");
        }
        else if (!strcmp(cmd, "pop")) {
            printf(">> enter the key to pop: ");
            scanf("%d", &key);
            root = root->pop(key);
            printf(">> ");
            root->traverse();
            printf(" \n");
        }
        else if (!strcmp(cmd, "traverse") || !(strcmp(cmd, "trv")))
        {
            printf(">> traversing the tree\n");
            printf("[");
            root->traverse();
            printf("]\n");
        }
        else if (!strcmp(cmd, "reset"))
        {
            printf(">> resetting the tree\n");
            root->reset();
            root = NULL;
            init = 1;
        }
        else if (!strcmp(cmd, "--help"))
        {
            printf("--------------------------------------------------------------------\n");
            printf("    exit                - exit the tree editing\n");
            printf("    traverse            - inorder traversal of the tree\n");
            printf("    reset               - reset the list\n");
            printf("    put                 - put a key into a tree\n");
            printf("    pop                 - pop a given key from the tree\n");
            printf("--------------------------------------------------------------------\n");
        }
        else
        {
            printf(">> no such command <%s>\n", cmd);
            printf(">> use --help to see the list of commands\n");
        }
    }
    return 0;
}

File: BST/tree.hs
import System.IO

data Tree = N | E Integer Tree Tree deriving(Show, Eq)
-- tree: null | int left right

prompt :: String -> IO ()
prompt str = (putStr str) >>= (\x -> hFlush stdout)

promptLn :: String -> IO ()
promptLn str = (putStrLn str) >>= (\x -> hFlush stdout)

t1 :: Tree
t1 = put (put (put (put (put (put (put (put (put (put N 5) 3) 8) 1) 4) 6) 1) 9) 7) 2

-- getKey: get the key of the tree
getKey :: Tree -> Integer
getKey N = -1
getKey (E x _ _) = x

showKey :: Integer -> String
showKey k
    | k < 0     = "(null)"
    | otherwise = show k

-- getLeft: get the left subtree
getLeft :: Tree -> Tree
getLeft N = N
getLeft (E _ t _) = t

-- getRight: get the right subtree
getRight :: Tree -> Tree
getRight N = N
getRight (E _ _ t) = t

-- contains: return true if the tree has the key, else return false
contains :: Tree -> Integer -> Bool
contains N _ = False
contains (E key lt rt) x 
    | key == x = True              -- search hit
    | key < x  = contains rt x     -- search the right tree
    | key > x  = contains lt x     -- search the left tree

-- put: put the value in the tree as a BST
put :: Tree -> Integer -> Tree
put N x = E x N N
put (E key lt rt) x
    | key == x = E key lt rt
    | key  < x = E key lt (put rt x)
    | key  > x = E key (put lt x) rt

-- lfm: leftmost node of the root
lfm :: Tree -> Tree
lfm (E key N rt) = (E key N rt) 
lfm (E key lt _) = lfm lt       -- recursive case
    
-- suc: find the successor of the root
suc :: Tree -> Tree
suc N               = N            -- null tree has no successor
suc (E key _ N)     = N            -- when there is no successor
suc (E key _ rt)    = lfm rt       -- leftmost of its subtree

-- pop: remove the node with the key = k
pop :: Tree -> Integer -> Tree
pop N _ = N
pop (E key lt rt) x
    | key == x = rep $ E key lt rt
    | key  < x = E key lt (pop rt x)
    | key  > x = E key (pop lt x) rt

-- rep: actual action of removing the node
rep :: Tree -> Tree
-- if the node has a single child, just connect it
rep (E key N rt)    = rt                          
rep (E key lt N)    = lt                        
-- if the node has both child, replace the value with the successor   
rep (E key lt rt)   = (E (getKey $ lfm rt) lt (rep $ lfm rt)) 

-- trv: inorder traversal of the tree
trv :: Tree -> String
trv N   = " empty tree "
trv t   = iot t

iot :: Tree -> String
iot N               = ""
iot (E key lt rt)   = (iot lt) ++ " " ++ (show key) ++ " " ++ (iot rt)

editTree :: Tree -> IO ()
editTree t = do
    prompt ">> enter a command: "
    cmd <- getLine
    case cmd of
        "put"       -> do
            prompt ">> enter an integer to put: "
            key <- getLine
            let k = read key :: Integer
            promptLn $ ">>" ++ (trv $ put t k)
            editTree $ put t k
        "pop"       -> do
            prompt ">> enter the key to pop: "
            key <- getLine
            let k = read key :: Integer
            promptLn $ ">>" ++ (trv $ pop t k)
            editTree $ pop t k
        "root"      -> do
            promptLn $ ">> the key of the root is " ++ (showKey $ getKey t)
            editTree t
        "traverse"  -> do
            promptLn ">> traversing the tree"
            promptLn $ "[" ++ (trv t) ++ "]"
            editTree t
        "reset"     -> do
            promptLn ">> resetting the tree..."
            editTree N
        "exit"      -> return()
        "--help"    -> do
            promptLn "--------------------------------------------------------------------"
            promptLn "    exit                - exit the tree editing"
            promptLn "    traverse            - inorder traversal of the tree "
            promptLn "    reset               - reset the list"
            promptLn "    root                - get the key of the root"
            promptLn "    put                 - put a key into a tree"
            promptLn "    pop                 - pop a given key from the tree"
            promptLn "--------------------------------------------------------------------"
            editTree t
        _           -> do
            promptLn $ ">> no such command <" ++ cmd ++ ">"
            promptLn ">> use --help to see the list of commands"
            editTree t


main :: IO ()
main = do
    promptLn "--------------------------------------------------------------------"
    promptLn "Binary Search Tree in Haskell"
    editTree N
    promptLn "--------------------------------------------------------------------"
